### ISSUE TITLE
CsString & utf8

### ISSUE THREAD

**Issue by VikingExplorer** [copperspice/copperspice#195]
Now that c++20 is here, are you anticipating that the design of CsString would be updated to account for std::u8string?

If not, what advantages would CsString have over std::u8string?

Thanks in advance!

**Comment by agserm** [comment]
Interesting question. This is a using declaration and it is something anyone could have added as of C++20 because of the addition of char8_t. The char8_t data type is guaranteed to be equivalent to unsigned char but it is a unique type intended to only contain UTF-8 data.

It's important to be aware the new string class does not contain any functionality to handle encodings. The SG16 study group has been working on a proposal for an encoding-aware class called std::text. However, the general opinion is this proposal is not likely to be part of the standard until at least C++29.

CsString is is a standalone library, independent of CopperSpice, that natively supports encoding and decoding utf-8 and utf-16 text. It can be used as a drop-in replacement for std::string. If you have other questions, please let us know.

**Comment by VikingExplorer** [comment]
My current String class encapsulates a std::wstring. I'm able to read unicode from the database, store it internally, and display it on the screen.  After reading the utf8 manifesto, I feel like I should refactor. There is a lot of converting to/from narrow/wide that I should try to avoid. 

I could replace std::wstring with std::u8string OR with CsBasicString<utf8>.

From watching the YouTube video on CsString, I vaguely remember CsString functionality relating to searching & replacing text?

By encoding & decoding, I assume you are referring to converting to/from unicode?  If so, I would have very little use for that functionality.

**Comment by VikingExplorer** [comment]
Maybe I'm wrong about that.  If a user somewhere on earth opened a file using my software. Could I have a problem if it's not Unicode?

**Comment by agserm** [comment]
This is a great conversation and you have piqued my interest. I want to first make sure we are using the same terminology.

You mentioned you "read unicode". What you read were bytes which represent symbols found in the Unicode table. Each symbol has a corresponding number, and these numbers are referred to as "codepoints". These codepoints are stored using some encoding such as UCS-4, UTF-8, UTF-16, etc. When you read data you must know what encoding it is using or you will interpret some of the codepoints incorrectly. Since std::wstring stores 16-bit values on Windows and 32-bit values on all other platforms, it is error prone to use in cross-platform code.

How does your String class handle encodings? Remember, simply reading data into std::wstring or std::u8string does not guarantee a specific encoding. How do you handle surrogate pairs? We suggest moving to CsString since it is well tested code which handles the complexities of UTF-8 and UTF-16 encoding.

**Comment by agserm** [comment]
Your question about a file that "is not Unicode" is mixing up terms. If you expected their file to be in UTF-8, and it is UTF-16 you have an encoding problem. Both UTF-8 and UTF-16 contain Unicode symbols.

**Comment by VikingExplorer** [comment]
I take your point regarding the difference between codepoints & encoding.  I've read various detailed explanations about that over the years.

However, I'm still left wondering about how a problem actually occurs.  XML files should have the encoding specified.

Windows 10 provides a configuration setting for non-Unicode programs, but this wouldn't apply to our software. It seems like Windows 10 now defaults to UTF8.  Code pages seem obsolete. Linux & macOS already have the native encoding as UTF-8.

Again, my choice is between std::u8string & CsBasicString.  What would be the pros & cons between these two?

**Comment by bgeller** [comment]
We did a video which gives a decent introduction to characters, codepoints, and encodings.  If you have not seen this you might want to watch this.  https://www.youtube.com/watch?v=nYzi0-VAXQM

An XML file defines a format for the text based on tags. The first line might say the file is UTF-8, but there is no character encoding. You are responsible for writing only UTF-8 encoded text to the file. The file is saved as raw bytes. 

Some of the terms you are mentioning are a bit confusing, I may have missed something in your post. There is no such thing as a Unicode program or a non-Unicode program. Windows, Linux, and OS X are not in control of how an application reads and writes files. Windows actually uses the UCS-2 encoding for file names but does not enforce anything about the contents of the file.  Can you let me know where you read that "code pages" are no longer used?

The **std::u8string** class added in C++20 does not handle encodings. This works nicely for ASCII characters which are the codepoints from 0 to 128. The Unicode codepoint for ö is U+00F6 which will take 2 bytes to store.  This is why UTF-8 strings are variable length.  The string "open" is 4 codepoints and takes 4 bytes to store.  The string "öffnen" is 6 codepoints and takes 7 bytes to store. The std::u8string class does not handle codepoints and will report the length of the second string as 7, which is not accurate.  There is no functionality in this class to use an iterator to walk a variable length encoded string. 

Our **CsBasicString** library automatically deals with variable length encodings internally. If you specify UTF-8 then for "öffnen"  size() returns 6 and size_storage() returns 7.  The iterators in our class work as you would expect, walking codepoint by codepoint.  In our opinion if you want to deal with UTF-8 encoded text our library is one of the only choices right now. 

Barbara

**Comment by VikingExplorer** [comment]
>> Unicode program or a non-Unicode program
![image](https://user-images.githubusercontent.com/15109208/115961192-4c536f80-a4e3-11eb-8c51-5cac88fbb7b5.png)

**Comment by VikingExplorer** [comment]
>> Unicode vs non-Unicode program

On Windows, vcxproj apps have this setting: https://docs.microsoft.com/en-us/cpp/text/support-for-unicode?view=msvc-160

This enables Unicode functionality, making it a "Unicode Program".  

2 bytes are enough to support characters for almost all modern languages
https://en.wikipedia.org/wiki/Plane_(Unicode)#Basic_Multilingual_Plane

Bytes 3 & 4 are used for rare or historical scripts.

I take your point about CsBasicString correctly reporting the length of a variable length encoded string like "öffnen".

However, it's fixing a problem caused by using a 1 byte char type.  Except for the weird edge case of an app that deals with historical scripts, std::wstring works well & reports the length of string correctly.

Maybe I'll do some performance testing to help me decide.

So, perhaps my true choice is between continuing to use wstring or refactor to use CsString.

Thank you so much for your responses.  It was really helpful.  Maybe this discussion will help other people as well.

**Comment by rwols** [comment]
"Unicode program" is a Microsoft concept from the 90s that is now misleading with the benefit of hindsight. All it means is that the program was compiled with the `UNICODE` macro defined, meaning that function-like macros like `SetWindowText` resolve to `SetWindowTextW` instead of `SetWindowTextA`.

The underlying encoding of `std::wstring` of modern Windows apps compiled with `UNICODE` is UTF-16.

You still seem to think that `std::string` and `std::wstring` somehow *carry* an encoding, but they **don't carry an encoding**. `std::string` and `std::wstring` are essentially sequences of bytes.

"Text" is a higher-level concept that can be seen as a tuple of (bytes, encoding). "Text" *does* carry an encoding, and knows its own encoding.

The initial question

> what advantages would CsString have over std::u8string?

is therefore comparing apples to oranges, as CsString is "text", while std::u8string is "sequence of bytes".

> perhaps my true choice is between continuing to use wstring or refactor to use CsString.

You're asking here: should I keep using "sequence of bytes", or "text"?

**Comment by VikingExplorer** [comment]
Your description of "Unicode program" is misleading.  It enables full Unicode functionality.  I provided an image of a current Windows setting.

"You still seem to think"

No, I don't.  Folks keep confusing the issue with minor implementation details.  I don't care that wstring isn't a self-aware AI, capable of knowing the encoding.  I'm working at a more abstract level.  It works in almost any reasonable situation.  Hate to tell you this, but the integer data type is also not aware of it's base 10 representation.

Big picture is that UTF is a UNICODE concept. The Unicode standard defines Unicode Transformation Formats (UTF) UTF-8, UTF-16, and UTF-32.

Barbara: "Can you let me know where you read that "code pages" are no longer used?

https://en.wikipedia.org/wiki/Windows_code_page
https://en.wikipedia.org/wiki/Code_page

Bottom line: there is ONLY UNICODE & it is the ONE GIANT CODE PAGE to rule them all

**Comment by VikingExplorer** [comment]
Example of your hyper-technicality:

"The underlying encoding of std::wstring of modern Windows apps compiled with UNICODE is UTF-16."
"You're asking here: should I keep using "sequence of bytes", or "text"?"

CsString & any other C++ class could be mischaracterized as just a sequence of bits.

**Comment by VikingExplorer** [comment]
![image](https://user-images.githubusercontent.com/15109208/115967624-a9f6b480-a501-11eb-8c1b-6bbcb526ff5b.png)

**Comment by bgeller** [comment]
@VikingExplorer ,  

I am sorry to be so blunt but you have some misconceptions about this topic and we really want to clear up your understandings of strings. The first point to know is that text is very complicated. Thinking otherwise is dangerous. 

> This enables Unicode functionality, making it a "Unicode Program"  . . .  It enables full Unicode functionality. I provided an image of a current Windows setting.

In regards to the MS image you posted this does not mean what you think. As @rwols mentioned, this is just a setting and in no way sets up or defines an encoding. There are a lot of very good articles which establish that MS made choices 20+ years ago and they got it wrong. They picked UCS-2 which predates UTF-16, both are 16 bit. At that time MS thought the Unicode table would never contain more than 64K worth of codepoints, they were dead wrong and they know it now.  Who knew how many emojis people would want to use and each one needs a unique codepoint. 

> 2 bytes are enough to support characters for almost all modern languages

This is completely wrong. The **Basic Multilingual Plane** or BMP covers the majority of codepoints used in everyday text, however it is simply not enough.  In UTF-16 you can use 16 bits to represent anything from 0 to 64k which is the BMP.  After that point you will need 32 bits for all other codepoints.  In UTF-8 you need 8 bits for 0 -128 and 16 bits for any codepoint from 129 to 32k.  It would be a total waste to use 16 bits for the codepoints from 0-128.  

The real issues start after the BMP where we have the **SMP** and then  **SIP**.  If you want to toss these out as useless then all musical notes, flags, mathematical notations, map symbols, and every single emoji would be gone.  In the SIP are the "less common CJK symbols". It turns out around 15% of all Japanese names require at least one codepoint from the SIP, which is stunning.

The Unicode consortium works like crazy to maintain a good standard. It is a ton of work and we are grateful for their service. 

> However, it's fixing a problem caused by using a 1 byte char type

What problem? Are you suggesting that UTF-8 be represented in two bytes for all codepoints less than 128?  This would make ASCII strings twice as long as needed. Silly and wasteful.  

I am not trying to be abrupt but you should go back and read the comments by @rwols again since it was extremely accurate and we are very impressed with his reply. ( Many thanks to Raoul. ) 

> Folks keep confusing the issue with minor implementation details. I don't care that wstring isn't a self-aware AI, capable of knowing the encoding

There is nothing minor about our implementation or the great work SG16 is trying to do for std::text. We are observers and commenters on SG16 and provide some input to them. Members of this study group have reviewed CsString as a possible solution to see how it could be adapted. 

> Big picture is that UTF is a UNICODE concept.

Nope, there is no actual "UTF" as an encoding.  Unicode is not an idea but a real standard that companies and computer languages recognize. The more common encodings the standard defines are UTF-8, UTF-16, and UTF-32. However there are  other encodings like UTF-16LE, UTF-16BE, UCS-2, UCS-4, BOCU-1, SCSU. 

The Unicode standard defines more than just encodings. There are rules for how to combine codepoints, how to display glyphs, using graphemes, pluralization (very messy in Irish), word breaking, and much more.  https://en.wikipedia.org/wiki/Unicode

> Barbara: "Can you let me know where you read that "code pages" are no longer used?

You provided links from the internet, but this does not mean they are accurate or true.  Looking at a current list of code pages 
you will see that Microsoft does in fact use them to handle Unicode. Code pages are assigned by Microsoft to associate a unique ID with an encoding.  For example, UTF-16LE is code page 1200 which UTF-8 is code page 65001. Maybe code pages are not a good idea, but they are fundamental part of the MS API.  Legacy can be annoying. 

> CsString & any other C++ class could be mischaracterized as just a sequence of bits.

CsString is actually a library which uses policy classes to support encodings. We currently have policies for UTF-8 and UTF-16. The CsString iterator class knows nothing about encodings and simply calls the policy class which understands "how to iterate".  If anyone wants to extend CsString for support for something like WTF-8 or MUTF-8, just add a set of new policies. No need to change the iterator class or anything else. Pretty nifty.  

@dascandy (Peter Bindels) did a really fun and informative talk at CppCon. He also provides input to SG16. 
https://www.youtube.com/watch?v=SMSmKg1nApM

Please understand we are all trying to help you understand this topic. Knowing that some really smart people on the C++ standards committee are struggling to work this out should let you know it is an on going issue.  We have talked with Tom Honermann, who is the author of std::u8string proposal and the head of SG16.  This new class did not solve encodings but it is a nice stepping stone.  

Barbara

**Comment by janwilmans** [comment]
Long story short, if you need text with an UTF-8 or UTF-16 encoding you can chose to use CsString. Personally, I don't deal with text a lot, and so I avoided character encodings like the plaque. That is until started writing unittests for text conversions, then CsString turned out to save the day.

**Comment by VikingExplorer** [comment]
@bgeller, I guess b/c your personality is overly detailed oriented, IMO, you're getting lost in technical details.

UCS-2: I'm sorry to be blunt, but you're using obsolete information. 
Reality: "Starting in Windows 10 build 17134 (April 2018 Update), the Universal C Runtime supports using a UTF-8 code page"
Reality: "Internally, Windows applications use the UTF-16 implementation of Unicode"

>>This is completely wrong. The Basic Multilingual Plane or BMP covers the majority of codepoints used in everyday text.

Which means that I'm completely RIGHT!!  LOL.  I hate to be blunt, but "UTF-16 is variable 2 or 4 bytes"

>> What problem? Are you suggesting that UTF-8 be represented in two bytes for all codepoints less than 128? This would make ASCII strings twice as long as needed. Silly and wasteful.

The problem you identified 15 hours ago as the main thing CsString fixes.   Calling UTF-16 "silly and wasteful" is quite a silly biased opinion. Only performance testing would determine if this optimization is worth it on a case-by-case basis.  Processing time vs memory usage.

>> Code Pages

Of course, legacy code pages exist, but the links I provided show that they are historical artifacts.  Windows 10 defaults to Unicode.  I've been converting files today using a handy tool (https://github.com/amrali-eg/EncodingChecker) and I can see that files I created a really long time ago are US-ASCII, while files from a long time ago are Windows-1252.  All new files are UTF-8.

An app that I would write would naturally default to reading customer files as Unicode. If the user was using a code page like Hebrew 1255, it would be their responsibility to convert that file.

>> CsString is actually a library which uses policy classes to support encodings.  ... WTF-8 or MUTF-8

Ahh, you've made the mistake of being too generic, over anticipating future needs that will never materialize. I'm an expert at generic code design & development. Generic code takes 3-20x more effort to build.  So, one needs to have at least 3 instances to make it worth it.  From this, CsString is perhaps like http://site.icu-project.org/home  OBE...

>> Please understand we are all trying to help you understand this topic

Please understand that virtue signaling meaningless obsolete technical details from the past is not persuasive.  From your LI profile, you appear to have only worked on Copper Spice.  You should consider that talking down to people with a lot more experience than you have reflects badly on you.  This should have ended when I closed the issue & thanked you.

**Comment by bgeller** [comment]
@VikingExplorer , 

Your last post was derogatory and simply uncalled for. This is a professional project and we expect people to be kind to one another and treat them with respect.  You have no idea who I am and should never comment on me as a person. Please edit your comment and remove all of the personal attacks.

**Comment by VikingExplorer** [comment]
I removed the only personal reference I found.   You don't appear to know how much experience I have.  Respect is earned.  You seem oblivious to how condescending you appear to be.

**Comment by janwilmans** [comment]
@VikingExplorer who are you?

**Comment by bgeller** [comment]
This is your second "CoC violation" warning. We have just received notice from 3 other people who are offended with your recent post.  Any further messages which are not on topic and professional will be deleted. Respect is indeed earned and insulting people reduces the respect others will have for you.
